File: python/examples/basic.py
"""Minimal end-to-end example. Mirrors js/examples/basic.ts.

Run from a venv where the package is installed (`uv pip install -e .`):
    python examples/basic.py

Uses the shared sample fixtures in ../../examples/. Requires `npx` on PATH for
the `@modelcontextprotocol/server-everything` stdio server.
"""
from __future__ import annotations

import asyncio
import json
import os

from toolnexus import create_toolkit

_HERE = os.path.dirname(os.path.abspath(__file__))
_EXAMPLES = os.path.normpath(os.path.join(_HERE, "..", "..", "examples"))


async def main() -> None:
    tk = await create_toolkit(
        mcp_config=os.path.join(_EXAMPLES, "mcp.json"),
        skills_dir=os.path.join(_EXAMPLES, "skills"),
    )

    print("MCP status:", tk.mcp_status())
    print("Tools:", [f"{t.name} ({t.source})" for t in tk.tools()])
    print("\nSystem-prompt skill catalog:\n" + tk.skills_prompt())

    print("\nOpenAI tool schema (first 2):")
    print(json.dumps(tk.to_openai()[:2], indent=2))

    # Load a skill (progressive disclosure) — works with no MCP servers running.
    res = await tk.execute("skill", {"name": "hello-world"})
    print("\nskill(hello-world) ->\n" + res.output)

    await tk.close()


if __name__ == "__main__":
    asyncio.run(main())

File: elixir/examples/basic.exs
# Minimal end-to-end example. Mirrors js/examples/basic.ts and python/examples/basic.py.
#
# Run from the elixir/ directory (deps fetched):
#     mix run examples/basic.exs
#
# Uses the shared sample fixtures in ../examples/. The `everything` MCP server in
# that mcp.json needs `npx` on PATH; if it is missing the server is isolated
# (status "failed") and the rest still runs — the skill path needs no network.

here = Path.dirname(__ENV__.file)
examples = Path.expand(Path.join([here, "..", "..", "examples"]))

{:ok, tk} =
  Toolnexus.create_toolkit(
    mcp_config: Path.join(examples, "mcp.json"),
    skills_dir: Path.join(examples, "skills")
  )

IO.puts("MCP status: #{inspect(Toolnexus.Toolkit.mcp_status(tk))}")

IO.puts(
  "Tools: " <>
    inspect(Enum.map(Toolnexus.Toolkit.tools(tk), &"#{&1.name} (#{&1.source})"))
)

IO.puts("\nSystem-prompt skill catalog:\n" <> Toolnexus.Toolkit.skills_prompt(tk))

IO.puts("\nOpenAI tool schema (first 2):")
IO.puts(Jason.encode!(Enum.take(Toolnexus.Toolkit.to_openai(tk), 2), pretty: true))

# Load a skill (progressive disclosure) — works with no MCP servers running.
res = Toolnexus.Toolkit.execute(tk, "skill", %{"name" => "hello-world"})
IO.puts("\nskill(hello-world) ->\n" <> res.output)

Toolnexus.Toolkit.close(tk)
